Smt/Graph.lean

```diff
@@ -20,62 +18,43 @@ variable {α} {β} [BEq α] [Hashable α] (g : Graph α β) (v u : α) (e : β)
 
 def empty : Graph α β := HashMap.empty
 
-def vertices : List α := g.fold (λ a v _ => v :: a) []
+def vertices : List α := g.fold (fun a v _ => v :: a) []
+
+def neighbors? : Option (List α) :=
+  g.find? v >>= fun es => some (es.fold (fun a v _ => v :: a) [])
 
-def neighbors : Option (List α) :=
-  (g.find? v).bind λ es => some (es.fold (λ a v _ => v :: a) [])
+def neighbors! : List α := match (g.neighbors? v) with
+  | some ns => ns
+  | none    => panic! "vertex is not in the graph"
 
 def addVertex : Graph α β := g.insert v HashMap.empty
 
 def addEdge : Graph α β := g.insert v ((g.find! v).insert u e)
 
-def weight : Option β := (g.find? v).bind λ es => es.find? u
+def weight? : Option β := g.find? v >>= fun es => es.find? u
 
 partial def dfs [Monad m] (f : α → m Unit) : m Unit :=
   StateT.run' (s := HashSet.empty) do
     for v in g.vertices do
       visitVertex v
-  where
-    visitVertex (v : α) : StateT (HashSet α) m Unit := do
-      let vs ← get
-      if vs.contains v then
-        return ()
-      set (vs.insert v)
-      match g.neighbors v with
-      | none    => f v
-      | some ns =>
-        for u in ns do
-          visitVertex u
-        f v
-
-def formatGraph [ToFormat α] [ToFormat β] : Format :=
-  Format.text "{" ++ Format.joinSep (g.vertices.map format') ","
-                  ++ Format.text "}"
-  where
-    format' (v : α) : Format :=
-     Format.text "(" ++ format v ++ Format.text ":"
-                     ++ format'' (g.neighbors v) ++ Format.text ")"
-    format'' : Option (List α) → Format
-      | none    => Format.nil
-      | some es => Format.joinSep es ","
-
-instance [ToFormat α] [ToFormat β] : ToFormat (Graph α β) where
-  format (g) := Graph.formatGraph g
-
-open Lean in
-def toMessageData [ToMessageData α] [ToMessageData β] : MessageData :=
```

Comment:
This instance makes the graph vertices in traces clickable when they are `Expr`s, so it'd be better to keep it.

Reply:
Huh... Been using this feature for a while and never knew how it worked! I originally removed it because the `node` constructor got replaced with `trace`. I added it back.